### PR TITLE
Issues/3367 introduce autoclose

### PR DESCRIPTION
**Pull request by ibrahimesseddyq**
## Overview

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AutoClose.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+@Target(ElementType.FIELD)
+@API(status = API.Status.EXPERIMENTAL, since = "5.9")
+public @interface AutoClose {
+	/**
+	 * Specifies the name of the method to invoke for closing the resource.
+	 * The default value is "close".
+	 *
+	 * @return the method name for closing the resource
+	 */
+	String value() default "close";
```
Comment (bjmi) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/AutoClose.java:53]: - If it is allowed to specify a different method name then the annotated resource does not necessarily have to implement `java.io.Closeable`.
- `value()` isn't evaluated and so a different method other than `close()` isn't invoked.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 2]
```diff
@@ -45,6 +45,7 @@ in the `junit-jupiter-api` module.
 | `@ExtendWith`            | Used to <<extensions-registration-declarative,register extensions declaratively>>. Such annotations are _inherited_.
 | `@RegisterExtension`     | Used to <<extensions-registration-programmatic,register extensions programmatically>> via fields. Such fields are _inherited_ unless they are _shadowed_.
 | `@TempDir`               | Used to supply a <<writing-tests-built-in-extensions-TempDirectory,temporary directory>> via field injection or parameter injection in a lifecycle method or test method; located in the `org.junit.jupiter.api.io` package.
+| @AutoClose | Indicates that a field in a JUnit 5 test class represents a resource that should be automatically closed after test execution. The field must implement the java.io.Closeable interface. The class must be annotated with `@ExtendWith(AutoCloseExtension.class)`.
```
Comment (bjmi) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:48]: It should work out of the box like `@TempDir` without specifying `@ExtendWith(AutoCloseExtension.class)` explicitly.  
As `@AutoClose` is an opt-in feature it should not interfere with other extension if this  extension isn't used but registered.

**Comment by bjmi** [comment]
A different approach would 

1. find all `@AutoClose` annotated fields,
2. wraps them in ExtensionContext.Store.[CloseableResource](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/ExtensionContext.Store.CloseableResource.html)
3. put them to the provided ExtensionContext.[Store](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/ExtensionContext.Store.html)

and let the JUnit engine close the resources when the Store is closed.

**Comment by marcphilipp** [comment]
@ibrahimesseddyq Do you have time to adjust your PR according to the team decision in https://github.com/junit-team/junit5/issues/3367#issuecomment-1835939613?

**Comment by ibrahimesseddyq** [comment]
@marcphilipp i will be working on the changes from today

**Comment by sormuras** [comment]
Touching over 800 files with `+86,253 −2` changes does not sound right.

**Comment by ibrahimesseddyq** [comment]
> Touching over 800 files with `+86,253 −2` changes does not sound right.

didn't commit gradle folder , dont know why that happened , i will correct that

**Comment by ibrahimesseddyq** [comment]
refactored code to functionnal style(all extensions are using this style) & added context.Store
 last time :checkstyleNohttp still here(caused by gradle folder)

**Comment by ibrahimesseddyq** [comment]
@sormuras @marcphilipp @bjmi some ideas to fix that gradle folder problem?

**Comment by ibrahimesseddyq** [comment]
> refactored code to functionnal style(all extensions are using this style) & added context.Store ,last time error :checkstyleNohttp still here(caused by gradle folder)

**Comment by marcphilipp** [comment]
I'd recommend rebasing and changing the commit that added the gradle folder. However, there's another PR for this issue by @bjmi now so we should first decide for one before making you do additional work.

**Comment by sbrannen** [comment]
Thanks again for the PR, @ibrahimesseddyq! 👍 

Please note, however, that this has now been:

- superseded by #3592

**Pull request closed without merging**